streaming/index.js

```diff
@@ -348,28 +364,40 @@ const startServer = async () => {
   const isInScope = (req, necessaryScopes) =>
     req.scopes.some(scope => necessaryScopes.includes(scope));
 
+  const ACCESS_TOKEN_UPDATE_FREQUENCY = 24 * 60 * 60 * 1000;
+
   /**
+   * Fetches the account from the access token, updating access token usage if necessary.
+   * `req.ip` comes from `proxyaddr`
    * @param {string} token
-   * @param {any} req
+   * @param {http.IncomingMessage & { ip: string } & ResolvedAccount} req
    * @returns {Promise<ResolvedAccount>}
    */
   const accountFromToken = async (token, req) => {
-    const result = await pgPool.query('SELECT oauth_access_tokens.id, oauth_access_tokens.resource_owner_id, users.account_id, users.chosen_languages, oauth_access_tokens.scopes FROM oauth_access_tokens INNER JOIN users ON oauth_access_tokens.resource_owner_id = users.id WHERE oauth_access_tokens.token = $1 AND oauth_access_tokens.revoked_at IS NULL LIMIT 1', [token]);
+    const result = await pgPool.query('SELECT oauth_access_tokens.id, oauth_access_tokens.resource_owner_id, users.account_id, users.chosen_languages, oauth_access_tokens.scopes, oauth_access_tokens.last_used_at FROM oauth_access_tokens INNER JOIN users ON oauth_access_tokens.resource_owner_id = users.id WHERE oauth_access_tokens.token = $1 AND oauth_access_tokens.revoked_at IS NULL LIMIT 1', [token]);
 
-    if (result.rows.length === 0) {
+    if (result.rows.length === 0 || result.rows.length > 1) {
       throw new AuthenticationError('Invalid access token');
     }
 
-    req.accessTokenId = result.rows[0].id;
-    req.scopes = result.rows[0].scopes.split(' ');
-    req.accountId = result.rows[0].account_id;
-    req.chosenLanguages = result.rows[0].chosen_languages;
+    const accessToken = result.rows[0];
+
+    // Track the usage of the access token if necessary:
+    // This is the same code as: app/controllers/concerns/api/access_token_tracking_concern.rb
+    if (accessToken.last_used_at === null || accessToken.last_used_at < Date.now() - ACCESS_TOKEN_UPDATE_FREQUENCY) {
+      await pgPool.query('UPDATE "oauth_access_tokens" SET "last_used_at" = $2, "last_used_ip" = $3 WHERE "oauth_access_tokens"."id" = $1', [ accessToken.id, new Date(), req.ip ]);
```

Comment:
I haven't been able to test this if req.ip returns say null; so it may be worth deploying this in a production environment to check that it all works as expected.

Reply:
I would wrap this into a `try … catch` block, if this fails we should not crash everything?

Also I dont like having the streaming server make a write query to the database, especially as we would like to decouple the streaming server from the DB at some point. I assume we dont have an easy way for the streaming server to make a query to the main app that would update this field?

Reply:
Correct, we don't really have a way to do this without doing a request back to the main app at this time, and we don't have anything that'd be an appropriate API call for that.

Reply:
Happy to wrap in a try/catch and just log the error if one occurs.

Reply:
Have wrapped in try / catch and supported the case where `req.ip` is not a valid IP address.